gdnative-derive/src/export_enum.rs

```diff
@@ -0,0 +1,152 @@
+    let impl_block = quote! {
+        impl ::gdnative::core_types::ToVariant for #enum_ty {
+            #[inline]
+            fn to_variant(&self) -> ::gdnative::core_types::Variant {
+                (*self as i64).to_variant()
+            }
+        }
+    };
+
+    Ok(impl_block)
+}
+
+fn impl_from_variant(enum_ty: &syn::Ident, data: &syn::DataEnum) -> syn::Result<TokenStream2> {
+    let n = quote! { n };
+    let arms = data
+        .variants
+        .iter()
+        .map(|variant| {
+            let ident = &variant.ident;
+            if !matches!(variant.fields, syn::Fields::Unit) {
+                Err(syn::Error::new(
+                    ident.span(),
+                    "#[derive(ExportEnum)] only support unit variant",
+                ))
+            } else {
+                Ok(quote! {
+                    if #n == #enum_ty::#ident as i64 { Ok(#enum_ty::#ident) }
+                })
+            }
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+    let expected_variants = data
+        .variants
+        .iter()
+        .map(|variant| quote! { stringify!(#variant.ident) });
+
+    let impl_block = quote! {
+        impl ::gdnative::core_types::FromVariant for #enum_ty {
+            #[inline]
+            fn from_variant(variant: &::gdnative::core_types::Variant) -> Result<Self, ::gdnative::core_types::FromVariantError> {
+                let #n = variant.try_to::<i64>()?;
+                #(#arms)else *
+                else {
+                    Err(FromVariantError::UnknownEnumVariant {
+                        variant: variant.to_string(),
+                        expected: &[#(#expected_variants),*],
+                    })
+                }
+            }
+        }
+    };
+
+    Ok(impl_block)
+}
+
+fn impl_export(enum_ty: &syn::Ident, data: &syn::DataEnum) -> syn::Result<TokenStream2> {
+    let mappings = data.variants.iter().map(|variant| {
+        let key = &variant.ident;
+        let val = quote! { #enum_ty::#key as i64 };
+        quote! { (stringify!(#key).to_string(), #val) }
+    });
+    let impl_block = quote! {
+        impl ::gdnative::export::Export for #enum_ty {
+            type Hint = ::gdnative::export::hint::IntHint<i64>;
+            #[inline]
+            fn export_info(hint: Option<Self::Hint>) -> ::gdnative::export::ExportInfo {
+                if let Some(hint) = hint {
+                    return hint.export_info();
+                } else {
+                    let mappings = vec![ #(#mappings),* ];
+                    let enum_hint = ::gdnative::export::hint::EnumHint::with_numbers(mappings);
+                    return ::gdnative::export::hint::IntHint::<i64>::Enum(enum_hint).export_info();
+                }
+            }
+        }
+    };
+
+    Ok(impl_block)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn deny_non_unit_variant() -> syn::Result<()> {
```

Comment:
Tests for proc-macros are better modeled as [UI tests](https://github.com/godot-rust/gdnative/tree/master/gdnative/tests/ui), as they not only allow us to check if the behavior is correct (with a real compiler, too), but also whether the error messages we show to users are up to par.

Reply:
Done